### PR TITLE
Feat 1863

### PR DESCRIPTION
**Pull request by marharyta**
# Definition of Done / Review checklist

## Reviewability
- [x] link to issue
- [ ] note if PR is on top of other PR
- [ ] note if related change in rems-deploy repo
- [x] consider adding screenshots for ease of review
Screenshot 2020-09-08 at 10 32 05

## Documentation
- [x] update changelog if necessary
- [x] add or update docstrings for namespaces and functions
- [x] components are added to guide page
- [ ] documentation _at least_ for config options (i.e. docs folder)
- [ ] ADR for major architectural decisions or experiments

## Different installations
- [ ] new configuration options added to rems-deploy repository
- [ ] instance specific translations (i.e. LBR kielivara)

## Testing
- [ ] complex logic is unit tested
- [ ] valuable features are integration / browser / acceptance tested automatically

## Accessibility
- [ ] all icons have the aria-label attribute
- [ ] all fields have a label
- [ ] errors are linked to fields with aria-describedby
- [ ] contrast is checked
- [ ] conscious decision about where to move focus after an action

## Follow-up
- [ ] new tasks are created for pending or remaining tasks
- [x] no critical TODOs left to implement

### CODE REVIEW

File: src/cljs/rems/fields.cljs [hunk 1]
```diff
@@ -15,6 +15,19 @@
 (defn field-name [field]
   (str "form-" (getx field :form/id) "-field-" (getx field :field/id)))
 
+(defn collapse [info-id info-text focus-when-collapse-opened body-text]
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs]: Should add docstring and use keyword arguments here.

File: src/cljs/rems/fields.cljs [hunk 2]
```diff
@@ -142,24 +142,30 @@
...
-                           :required (not optional)
-                           :aria-invalid (when validation true)
-                           :aria-describedby (when validation
-                                               (str (field-name opts) "-error"))
-                           :max-length max-length
-                           :class (when validation "is-invalid")
-                           :value value
-                           :on-change (comp on-change event-value)}]]))
+     [:<>
+      (if collapse-info [collapse
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs:152]: The proper construct to use is `when` if you don't have the else clause of an `if`.

File: src/cljs/rems/fields.cljs [hunk 3]
```diff
@@ -142,24 +142,30 @@
   (.. event -target -value))
 
 (defn text-field
-  [{:keys [validation on-change] :as opts}]
+  [{:keys [validation on-change collapse-info info-text] :as opts}]
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs:145]: There should be no need to have both `collapse-info` and `info-text`. Collapse can be hidden if there is no `info-text` given?

File: src/cljs/rems/fields.cljs [hunk 4]
```diff
@@ -142,24 +142,30 @@
...
-                           :max-length max-length
-                           :class (when validation "is-invalid")
-                           :value value
-                           :on-change (comp on-change event-value)}]]))
+     [:<>
+      (if collapse-info [collapse
+                         (field-name opts)
+                         (str "some text")
+                         focus-when-collapse-opened
+                         [:span (str info-text)]])
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs:156]: There is no need of `str` here if you implement then `if/when` without having a separate `collapse-info`.

File: src/cljs/rems/fields.cljs [hunk 5]
```diff
@@ -142,24 +142,30 @@
   (.. event -target -value))
 
 (defn text-field
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs:144]: I think it would be really good to add the docstring with all the arguments here.

File: src/cljs/rems/fields.cljs [hunk 6]
```diff
@@ -401,6 +407,12 @@
                     :field/type :text
                     :field/title {:en "Title"}
                     :field/placeholder {:en "placeholder"}}])
+   (example "field of type \"text\" with info field "
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs:410]: Isn't this example missing the `info-text`?

**Review by Macroz — CHANGES_REQUESTED** [review]

**Comment by marharyta** [comment]
Screenshot 2020-09-03 at 14 35 02
I am rather confused as of why when statement does not work here.

File: src/cljs/rems/fields.cljs [hunk 7]
```diff
@@ -129,24 +142,30 @@
   (.. event -target -value))
 
 (defn text-field
-  [{:keys [validation on-change] :as opts}]
+  [{:keys [validation on-change collapse-info-btn collapse-aria-lebel info-text] :as opts}]
```
Comment (opqdonut) [inline src/cljs/rems/fields.cljs:145]: lebel should probably be label

Comment (marharyta) [reply src/cljs/rems/fields.cljs:145]: yes, typo!

**Comment by marharyta** [comment]
Sep-03-2020 14-51-10

noticed a small bug after adding to guide, debugging

File: src/cljs/rems/fields.cljs (hunk 4 above)
Comment (Macroz) [reply src/cljs/rems/fields.cljs]: Also no need for the `:span` either. One could give a more complicated info-text and these are preventing it.

Comment (marharyta) [reply src/cljs/rems/fields.cljs]: ok, I see your point

File: src/cljs/rems/fields.cljs (hunk 2 above)
Comment (marharyta) [reply src/cljs/rems/fields.cljs]: Yes, I realized it shortly after and fixed that, but thanks for commentig

File: src/cljs/rems/fields.cljs (hunk 5 above)
Comment (marharyta) [reply src/cljs/rems/fields.cljs:144]: Maybe a better option would be adding it to ``` field-wrapper```?

**Comment by marharyta** [comment]
Sep-03-2020 17-18-06
As you can see, the problem with the animation on click is addressed, but the problem with "scroll glitch" is not yet

File: src/cljs/rems/fields.cljs (hunk 5 above)
Comment (Macroz) [reply src/cljs/rems/fields.cljs:152]: Yes, also I think all the field types should support the info, not just basic text field.

**Comment by marharyta** [comment]
I actually realised I ahve to add aria lebel and aria lebelby, I ll do it next

File: src/cljs/rems/fields.cljs [hunk 8]
```diff
@@ -375,25 +393,33 @@
...
                     :field/type :text
                     :field/title {:en "Title"}
                     :field/placeholder {:en "placeholder"}}])
+   (example "field of type \"text\" with info field "
+            [field {:form/id 1
+                    :field/id "1"
+                    :field/type :text
+                    :field/title {:en "Title"}
+                    :field/placeholder {:en "placeholder"}
+                    :field/collapse-aria-label {:en "Collapse aria lebel"}
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs:421]: typo

**Review by Macroz — CHANGES_REQUESTED** [review]

File: src/cljs/rems/fields.cljs (hunk 8 above)
Comment (marharyta) [reply src/cljs/rems/fields.cljs]: Fixed

File: src/cljs/rems/fields.cljs [hunk 9]
```diff
@@ -59,6 +73,8 @@
   :field/optional - boolean, true if the field is not required
   :field/value - string, the current value of the field
   :field/previous-value - string, the previously submitted value of the field
+  :field/info-text - text for collapsable info field
+  :field/collapse-aria-label - text for aria-label of info button
```
[the remark this answers is not in the document]
Comment (marharyta) [reply src/cljs/rems/fields.cljs:77]: I addressed that in my code 
But I agree, maybe I can refactor

File: src/cljc/rems/common/form.cljc [hunk 10]
```diff
@@ -26,6 +26,9 @@
 (defn supports-visibility? [field]
   true) ; at the moment all field types
 
+(defn supports-collapsable-info-text? [field]
+  (contains? #{:text :texta} (:field/type field)))
+
```
[the remark this answers is not in the document]
Comment (marharyta) [reply src/cljc/rems/common/form.cljc:31]: Because this was in the initial specs in the task. But we can add it to all.

File: src/clj/rems/api/schema.clj [hunk 11]
```diff
@@ -220,6 +223,7 @@
 (s/defschema FormTemplateOverview
   (dissoc FormTemplate :form/fields))
 
+;;; instance for form template once filled in by user
```
[the remark this answers is not in the document]
Comment (marharyta) [reply src/clj/rems/api/schema.clj:226]: Me and Joel thought comments like this could be pushed to add some clarity

File: src/clj/rems/db/test_data.clj [hunk 12]
```diff
@@ -22,6 +22,11 @@
   (:import [java.util UUID]
            [java.util.concurrent Executors Future]))
 
+;;; do not forget lein run reset, lein run test-data
+;;; fix schema errors 
+;;; fix create_form
+;;; 
+
```
[the remark this answers is not in the document]
Comment (marharyta) [reply src/clj/rems/db/test_data.clj:29]: aaand this one should be removed, by bad

File: src/clj/rems/layout.clj [hunk 13]
```diff
@@ -35,7 +35,11 @@ window.rems = {
                      #(text :t.header/title))]
            (include-css "/assets/bootstrap/css/bootstrap.min.css")
            (include-css "/assets/font-awesome/css/all.css")
-           (include-css (css-filename (env :default-language)))]
+           (include-css (css-filename (env :default-language)))
+           [:style "body {scroll-behavior: smooth;}
+                  @media screen and (prefers-reduced-motion: reduce) {
+                     body {scroll-behavior: auto;
+                    }} "]]
```
[the remark this answers is not in the document]
Comment (marharyta) [reply src/clj/rems/layout.clj:42]: Yeas, good point.

File: src/clj/rems/api/schema.clj (hunk 11 above)
[the remark this answers is not in the document]
Comment (Macroz) [reply src/clj/rems/api/schema.clj:226]: This makes more sense than the others.

File: src/cljc/rems/common/form.cljc (hunk 10 above)
[the remark this answers is not in the document]
Comment (Macroz) [reply src/cljc/rems/common/form.cljc:31]: That task was not very well groomed.

File: src/cljs/rems/administration/create_form.cljs [hunk 14]
```diff
@@ -186,13 +188,13 @@
...
-                                               (if edit?
-                                                 (::form-id db)
-                                                 (response :id))))))
+                              :top
+                              description
+                              (fn [response]
+                                (navigate! (str "/administration/forms/"
+                                                (if edit?
+                                                  (::form-id db)
+                                                  (response :id))))))
```
[the remark this answers is not in the document]
Comment (marharyta) [reply src/cljs/rems/administration/create_form.cljs:196]: This is something that we have already discussed in May, I believe. My editor automatically edits these, sometimes it is hard for me to keep track. Is there a git hook or something automatic we can use to solve this problem once and for all? I am willing to do that PR, I just do not think it is productive for either me or you to watch for every small indentation change

File: src/cljc/rems/common/form.cljc (hunk 10 above)
[the remark this answers is not in the document]
Comment (marharyta) [reply src/cljc/rems/common/form.cljc]: ok.

File: src/cljs/rems/administration/create_form.cljs (hunk 14 above)
[the remark this answers is not in the document]
Comment (opqdonut) [reply src/cljs/rems/administration/create_form.cljs:196]: I think I promised to configure clj-format for REMS in the spring. I'll do it now.

**Comment by marharyta** [comment]
I believe @Macroz s comment about styles have been addressed

File: src/cljs/rems/administration/create_form.cljs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Macroz) [reply src/cljs/rems/administration/create_form.cljs:196]: Well one thing to make sure is that you always read the diff and make sure you commit only what you intended. I.e. as a start I would not commit changes to an unrelated part in the same commit but a separate format commit. For clarity I would perhaps avoid doing any indentation changes for a start until I get the hang of it. Configuring the IDE to match the project settings is something to look at. cljfmt should help too. Controlling the content of the PR is an important skill to get your PR reviewed in most open-source projects so I would spend time to learn that.

File: CHANGELOG.md [hunk 15]
```diff
@@ -51,6 +51,8 @@ Changes since v2.12
 - The OIDC attribute to use as the rems userid is now configurable via the `:oidc-userid-attribute`. See [docs/configuration.md](docs/configuration.md). (#2281)
 - The `:oidc-additional-authorization-parameters` config option. See [config-defaults.edn](resources/config-defaults.edn)
 - Applicants can now permanently delete drafts. (#2219)
+- It is now possible to optionally add info icon to text fields with collapsable description text (#1863)
```
Comment (Macroz) [inline CHANGELOG.md:54]: How about rephrasing this like: `All fields can have an info text, shown if the small icon is clicked. (#1863)`

File: src/clj/rems/api/schema.clj [hunk 16]
```diff
@@ -175,6 +175,7 @@
 
 (def FieldId s/Str)
 
+;;; asking api for form templete
```
[the remark this answers is not in the document]
Comment (Macroz) [reply src/clj/rems/api/schema.clj:178]: To match the other comment, how about?
```
;;; template for a form field, before answering
```

File: src/cljs/rems/fields.cljs [hunk 17]
```diff
@@ -73,7 +88,9 @@
         optional (:field/optional opts)
         value (:field/value opts)
         previous-value (:field/previous-value opts)
-        max-length (:field/max-length opts)]
+        max-length (:field/max-length opts)
+        info-text (localized (:field/info-text opts))
+        collapse-aria-label (str "Info about " title)]
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs:93]: Like I said in another comment, I don't think there needs to be a parameter `collapse-aria-label` however this is now a hard-coded string. How about localizing it?

**Review by Macroz — CHANGES_REQUESTED** [review]
Found something that was not still solved and have a couple more suggestions on wordings.

File: resources/translations/fi.edn [hunk 18]
```diff
@@ -223,6 +223,7 @@
                 :field-n "Kenttä %1: %2"
                 :field-title "Kentän otsikko"
                 :field-type "Kentän tyyppi"
+                :info-text "Kenttätiedot, ylimääräinen kuvaus, esimerkkejä (valinnainen)"
```
Comment (Macroz) [inline resources/translations/fi.edn:226]: instead of `valinnainen` we use `ei pakollinen` in the others. This should be made to match. Also I think the long name is slightly confusing to users so maybe "Field description", "Kentän kuvaus", "Fält beskrivning"?

**Comment by Macroz** [comment]
I spent some time clicking in the UI and here's some further things that should be fixed here.
- [ ] The info text is in bold whereas only the field title should be. It should be in the regular weight.
- [ ] The info text does not preserve the spacing if you create a multi-paragraph description. Should have `white-space: pre-wrap` I think.
- [ ] Links in the info text are not rendered as links (they are rendered as links in the field title, have a look at that)
- Doesn't have to be fixed here, the preview of the form does not render the field names as bold so it looks different from the actual form filling.

What I typed in the form editor
in_editor

What is shown in the actual form
in_form

File: CHANGELOG.md (hunk 15 above)
Comment (marharyta) [reply CHANGELOG.md:54]: 👍

File: src/cljs/rems/fields.cljs (hunk 17 above)
Comment (marharyta) [reply src/cljs/rems/fields.cljs:93]: Yes, i wanted to do it, but forgot after the tests were failing, will address that

File: resources/translations/fi.edn (hunk 18 above)
Comment (marharyta) [reply resources/translations/fi.edn]: Addressed

**Comment by marharyta** [comment]
> I spent some time clicking in the UI and here's some further things that should be fixed here.
> 
> * [ ]  The info text is in bold whereas only the field title should be. It should be in the regular weight.
> * [ ]  The info text does not preserve the spacing if you create a multi-paragraph description. Should have `white-space: pre-wrap` I think.
> * [ ]  Links in the info text are not rendered as links (they are rendered as links in the field title, have a look at that)
> * Doesn't have to be fixed here, the preview of the form does not render the field names as bold so it looks different from the actual form filling.
> 
> What I typed in the form editor
> in_editor
> 
> What is shown in the actual form
> in_form

These are some great findings! I ll take a closer look at those issues!

**Comment by marharyta** [comment]
Regarding the @Macroz comment here
"The info text is in bold whereas only the field title should be. It should be in the regular weight."
- I do not see the text as bold in either one of those implementations

Screenshot 2020-09-10 at 11 25 49

File: test/clj/rems/test_browser.clj [hunk 19]
```diff
@@ -740,24 +746,29 @@
...
+
+
     (testing "fetch form via api"
       (let [form-id (Integer/parseInt (last (str/split (btu/get-url) #"/")))]
         (is (= {:form/id form-id
                 :organization {:organization/id "nbn" :organization/name {:fi "NBN" :en "NBN" :sv "NBN"} :organization/short-name {:fi "NBN" :en "NBN" :sv "NBN"}}
                 :form/title "Form editor test"
                 :form/fields [{:field/placeholder {:fi "" :en "" :sv ""}
                                :field/title {:fi "Description (FI)" :en "Description (EN)" :sv "Description (SV)"}
+                               :field/info-text {:en "", :fi "", :sv ""}
```
Comment (Macroz) [inline test/clj/rems/test_browser.clj:758]: Does this actually mean that one will always get an info-text even if the administrator left it empty? I think it would be cleanest not to return one at all if it was not set (since it's optional).

File: src/cljs/rems/fields.cljs [hunk 20]
```diff
@@ -375,25 +392,32 @@
...
                     :field/type :text
                     :field/title {:en "Title"}
                     :field/placeholder {:en "placeholder"}}])
+   (example "field of type \"text\" with info field"
+            [field {:form/id 1
+                    :field/id "1"
+                    :field/type :text
+                    :field/title {:en "Title"}
+                    :field/placeholder {:en "placeholder"}
+                    :field/info-text {:en "Extra information"}}])
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs:420]: This example could use line breaks and links to make it more representative.

File: src/cljs/rems/fields.cljs [hunk 21]
```diff
@@ -73,7 +88,9 @@
         optional (:field/optional opts)
         value (:field/value opts)
         previous-value (:field/previous-value opts)
-        max-length (:field/max-length opts)]
+        max-length (:field/max-length opts)
+        info-text (localized (:field/info-text opts))
+        collapse-aria-label (str (text :t.create-form/collapse-aria-label) (linkify raw-title))]
```
Comment (opqdonut) [inline src/cljs/rems/fields.cljs:93]: looks like you could just use `title` here instead of recomputing `(linkify raw-title)`

File: src/cljs/rems/fields.cljs [hunk 22]
```diff
@@ -271,19 +294,13 @@
...
-      [text :t.form/upload-extensions]
-      ": "
-      attachment-types/allowed-extensions-string]]))
+     [collapse
+      info-id
+      (text :t.form/upload-extensions)
+      focus-when-collapse-opened
+      [:span [text :t.form/upload-extensions]
+       ": "
+       attachment-types/allowed-extensions-string]]]))
```
Comment (opqdonut) [inline src/cljs/rems/fields.cljs:303]: Here you're passing a non-string `body-text` argument to `collapse`. Perhaps linkifying the body shouldn't be in `collapse` but instead the caller. We only use linkify for a couple of things but we use collapse everywhere.

File: src/clj/rems/css/styles.clj [hunk 23]
```diff
@@ -70,6 +69,8 @@
                              :font-weight 400
                              :font-style "normal"})))
 
+;; input.form-control white-space: pre-wrap
+
```
Comment (Macroz) [inline src/clj/rems/css/styles.clj:73]: remove commented code?

File: src/clj/rems/css/styles.clj [hunk 24]
```diff
@@ -653,8 +658,8 @@
    [:.form-group {:text-align "initial"}
     ;; make fieldset legends look the same as normal labels
     [:legend {:font-size "inherit"}]]
-   [:#application-fields
-    [:.application-field-label {:font-weight "bold"}]]
+   [:label.application-field-label {:font-weight "bold"}]
+   [:div.field-info {:font-weight "400"}]
```
Comment (Macroz) [inline src/clj/rems/css/styles.clj:662]: Above you have `[:.field-info {:white-space :pre-wrap}]` that you can combine here with `[:div.field-info {:font-weight "400"}]`.

File: src/cljs/rems/fields.cljs [hunk 25]
```diff
@@ -15,6 +15,20 @@
...
 
+(defn info-collapse [info-id aria-label-text focus-when-collapse-opened body-text]
+  [:<> [:button.info-button.btn.btn-link
+        {:data-toggle "collapse"
+         :href (str "#" (str info-id "-collapse"))
+         :aria-label aria-label-text
+         :aria-expanded "false"
+         :aria-controls (str info-id "-collapse")}
+        [:i.fa.fa-info-circle]]
+   [:div.field-info.collapse {:id (str info-id "-collapse")
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs:26]: The class should ideally match the component name i.e. `.info-collapse`. NB: you have to change the CSS selectors to match!

File: test/clj/rems/test_browser.clj [hunk 26]
```diff
@@ -842,24 +848,29 @@
         (btu/wait-page-loaded)
         (is (btu/visible? {:tag :label :class :application-field-label :fn/has-text "Option list (EN)"}))))
 
+
+
```
Comment (Macroz) [inline test/clj/rems/test_browser.clj:852]: Unnecessary change?

File: test/cljs/rems/administration/test_create_form.cljs [hunk 27]
```diff
@@ -171,6 +171,8 @@
                              :field/index 0
                              :field/title {:en "en title"
                                            :fi "fi title"}
+                             :field/info-text {:en "en info text"
+                                               :fi "fi info text"}
```
Comment (Macroz) [inline test/cljs/rems/administration/test_create_form.cljs:175]: I don't understand why all these have info-text defined. Surely it's not compulsory everywhere?

File: test/cljs/rems/test_fields.cljs [hunk 28]
```diff
@@ -20,9 +20,11 @@
...
-    (testing "previous value is same as current value"
-      (is (not (contains-hiccup? toggle-diff-button (field-wrapper {:form/id "foo" :field/id "foo" :field/value "foo" :field/previous-value "foo"} "<editor-component>")))))))
+     (testing "no previous value"
+        (is (not (contains-hiccup? toggle-diff-button 
+                                   (field-wrapper {:form/id "foo" :field/id "foo" :field/value "foo"} "<editor-component>")))))
+     (testing "has previous value"
+       (is (contains-hiccup? toggle-diff-button 
+                             (field-wrapper {:form/id "foo" :field/id "foo" :field/value "foo", :field/previous-value "bar"} "<editor-component>"))))
+     (testing "previous value is same as current value"
+       (is (not (contains-hiccup? toggle-diff-button (field-wrapper {:form/id "foo" :field/id "foo" :field/value "foo" :field/previous-value "foo"} "<editor-component>")))))))
```
Comment (Macroz) [inline test/cljs/rems/test_fields.cljs:30]: What is the change here?

**Comment by Macroz** [comment]
Ok now the result looks like it should. Still a bunch of minor issues to resolve. I think they are all visible here in the Conversation tab. One thing that came to my mind is that in this or next PR it would be nice to see the info-collapse tested in the browser tests.

File: test/cljs/rems/test_fields.cljs (hunk 28 above)
Comment (marharyta) [reply test/cljs/rems/test_fields.cljs:30]: hmm, no change really

File: test/cljs/rems/administration/test_create_form.cljs (hunk 27 above)
Comment (Macroz) [reply test/cljs/rems/administration/test_create_form.cljs:175]: My comment was about that there are now many places in the tests where you have added the same lines of `info-text`. Maybe some tests shouldn't have it and some should. After all it's not compulsory.

File: test/cljs/rems/test_fields.cljs (hunk 28 above)
Comment (Macroz) [reply test/cljs/rems/test_fields.cljs:30]: Ok just a formatting change I guess (split line + indent)

File: test/clj/rems/test_browser.clj (hunk 19 above)
Comment (marharyta) [reply test/clj/rems/test_browser.clj:858]: I am doing the last changes and it looks like tests fail without these fields

**Comment by marharyta** [comment]
Screenshot 2020-09-21 at 11 16 06

Once disable and trying to edit catalogue item, the form filed cannot be tweaked

**Comment by opqdonut** [comment]
Yeah that's probably bug https://github.com/CSCfi/rems/issues/2335

File: src/cljs/rems/fields.cljs [hunk 29]
```diff
@@ -97,7 +114,13 @@
       " "
       (if optional
         (text :t.form/optional)
-        (text :t.form/required))]
+        (text :t.form/required))
+      (when info-text
```
Comment (marharyta) [inline src/cljs/rems/fields.cljs:118]: @opqdonut what do you think?

Comment (opqdonut) [reply src/cljs/rems/fields.cljs:118]: It probably mostly works, but what about when a user fills in the info text in form editor but then clears it? I think you might end up with a `""` info text in that case.

Comment (marharyta) [reply src/cljs/rems/fields.cljs:118]: hmm, true. maybe I should do an empty string check?

Comment (opqdonut) [reply src/cljs/rems/fields.cljs:118]: yep! remember that info-text is a map here

File: src/cljs/rems/administration/create_form.cljs [hunk 30]
```diff
@@ -140,7 +140,10 @@
           :field/type (:field/type field)
           :field/optional (if (common-form/supports-optional? field)
                             (boolean (:field/optional field))
-                            false)}
+                            false)
+          }
+         (when (:field/info-text field)
```
Comment (marharyta) [inline src/cljs/rems/administration/create_form.cljs:145]: @opqdonut actually .. it is here ...

Comment (opqdonut) [reply src/cljs/rems/administration/create_form.cljs:145]: heh indeed, well my comments still stand :)

Comment (marharyta) [reply src/cljs/rems/administration/create_form.cljs:145]: 😸

File: src/cljs/rems/administration/create_form.cljs [hunk 31]
```diff
@@ -134,18 +134,21 @@
   (into {} (for [language languages]
              [language (trim-when-string (get lstr language ""))])))
 
+(defn- string-is-not-empty [s]
+  (when (string? s)
+    (not (clojure.string/blank? s))))
```
Comment (opqdonut) [inline src/cljs/rems/administration/create_form.cljs:139]: 1) `(clojure.string/blank? nil)` is `true` so no need for the `when` here
2) you can use `str/blank?` since we have a `(:require [clojure.string :as str])` in this namespace

Comment (marharyta) [reply src/cljs/rems/administration/create_form.cljs:139]: Good point!

File: src/cljs/rems/administration/create_form.cljs [hunk 32]
```diff
@@ -135,8 +135,10 @@
              [language (trim-when-string (get lstr language ""))])))
 
 (defn- string-is-not-empty [s]
-  (when (string? s)
-    (not (str/blank? s))))
+  (if (string? s)
+    (not (str/blank? s))
+    false
+    ))
```
Comment (opqdonut) [inline src/cljs/rems/administration/create_form.cljs:141]: as I mentioned earlier, you should be able to just define

```
(defn- string-is-not-empty [s]
  (not (str/blank? s)))
```

File: src/cljs/rems/administration/create_form.cljs [hunk 33]
```diff
@@ -134,13 +134,23 @@
...
 (defn- build-request-field [field languages]
   (merge {:field/id (:field/id field)
           :field/title (build-localized-string (:field/title field) languages)
           :field/type (:field/type field)
           :field/optional (if (common-form/supports-optional? field)
                             (boolean (:field/optional field))
                             false)}
+         (when (some true? 
+                (mapv (fn [l] (string-is-not-empty (l (:field/info-text field)))) 
+                   languages))
```
Comment (opqdonut) [inline src/cljs/rems/administration/create_form.cljs:152]: this can be simplified to:

```
(some (fn [l] (string-is-not-empty (l (:field/info-text field))))
      languages)
```

**Review by opqdonut — COMMENTED** [review]
some code golf suggestions

I think this is mergeable, what about you @Macroz ?

File: src/cljs/rems/administration/create_form.cljs (hunk 32 above)
Comment (Macroz) [reply src/cljs/rems/administration/create_form.cljs:141]: Also I think it's better to just not define that function and use not blank directly.

File: src/cljs/rems/administration/create_form.cljs (hunk 33 above)
Comment (Macroz) [reply src/cljs/rems/administration/create_form.cljs:152]: This should probably also rather be `not-any?` `str/blank?`?

**Comment by Macroz** [comment]
> some code golf suggestions
> 
> I think this is mergeable, what about you @Macroz ?

I would still like to see my points addressed (even the older ones that remain unresolved).

File: test/cljs/rems/administration/test_create_form.cljs (hunk 27 above)
Comment (opqdonut) [reply test/cljs/rems/administration/test_create_form.cljs:237]: now there's a test case both ways

File: src/cljs/rems/fields.cljs (hunk 25 above)
Comment (opqdonut) [reply src/cljs/rems/fields.cljs]: this seems done now

File: src/clj/rems/css/styles.clj (hunk 24 above)
Comment (opqdonut) [reply src/clj/rems/css/styles.clj]: this seems done now

**Comment by opqdonut** [comment]
I went through some of @Macroz's old comments and resolved the ones that are fixed.

File: src/cljs/rems/administration/create_form.cljs (hunk 32 above)
Comment (marharyta) [reply src/cljs/rems/administration/create_form.cljs:141]: Why  not define it?

Comment (Macroz) [reply src/cljs/rems/administration/create_form.cljs:141]: `str/blank?` is as good as standard function, no need to define one. Also `not a function` does not need a new name, especially so trivial as this one is. Also a later refactoring could remove the `not` entirely too.

File: src/cljs/rems/fields.cljs (hunk 1 above)
Comment (marharyta) [reply src/cljs/rems/fields.cljs]: > docstring

this is the only one unresolved I managed to find

File: src/cljs/rems/fields.cljs [hunk 34]
```diff
@@ -15,6 +15,20 @@
 (defn field-name [field]
   (str "form-" (getx field :form/id) "-field-" (getx field :field/id)))
 
+(defn collapse [info-id aria-label-text focus-when-collapse-opened body-text]
```
[the remark this answers is not in the document]
Comment (marharyta) [reply src/cljs/rems/fields.cljs]: another one.

File: src/cljs/rems/fields.cljs (hunk 9 above)
[the remark this answers is not in the document]
Comment (marharyta) [reply src/cljs/rems/fields.cljs]: I believe this one is resolved

File: src/cljs/rems/fields.cljs (hunk 25 above)
Comment (marharyta) [reply src/cljs/rems/fields.cljs]: resolved

File: test/cljs/rems/administration/test_create_form.cljs (hunk 27 above)
Comment (marharyta) [reply test/cljs/rems/administration/test_create_form.cljs:237]: Resolved

File: src/cljs/rems/fields.cljs (hunk 9 above)
[the remark this answers is not in the document]
Comment (opqdonut) [reply src/cljs/rems/fields.cljs]: yeah

File: src/cljs/rems/fields.cljs (hunk 1 above)
Comment (Macroz) [reply src/cljs/rems/fields.cljs]: Keyword arguments too.

File: src/cljs/rems/fields.cljs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Macroz) [reply src/cljs/rems/fields.cljs]: I originally meant that you could possibly remove the parameter altogether and just have "More info" as the text. Not 100% sure on the accessibility of it but it would remove the need to pass or maintain them.

[the remark this answers is not in the document]
Comment (marharyta) [reply src/cljs/rems/fields.cljs]: well, this is outdated, the only thing it has to do is to put in some code and we are already putting in a string here

[the remark this answers is not in the document]
Comment (Macroz) [reply src/cljs/rems/fields.cljs]: Yes I'm just wondering that could this have been left out completely and have always had the localized "More info". It's anyway marginally better to have "More info about foobar" in the text although `text-format` should be used and not string concatenation. We can leave it as it is.

File: src/cljs/rems/fields.cljs [hunk 35]
```diff
@@ -15,7 +15,13 @@
...
   (str "form-" (getx field :form/id) "-field-" (getx field :field/id)))
 
-(defn info-collapse [info-id aria-label-text focus-when-collapse-opened body-text]
+(defn info-collapse
+  "Bootstrap collapse field that shows extra information about input fields:
+   info-id - CSS id of the element being described
+   aria-label-text - text describing aria-label of collapse, see more https://developers.google.com/web/fundamentals/accessibility/semantics-aria/aria-labels-and-relationships
+   focus-when-collapse-opened - focus function that toggles the collapsable field
+   body-text - text that is shown as the info icon is toggles that described the field"
+  [{:keys [info-id aria-label-text focus-when-collapse-opened body-text]}]
```
Comment (Macroz) [inline src/cljs/rems/fields.cljs:24]: There are some tricks to writing docstrings we use like backquotes so here's an alternative phrasing more in line with the others.
```
  "Collapse field that shows extra information about input fields.

  `:info-id` - id of the element being described
  `:aria-label-text` - aria label for the info button
  `:focus-when-collapse-opened` - element that is focused when the info is opened
  `:body-text` - component that is shown if open"
```

**Review by Macroz — APPROVED** [review]
Have a look at the last comment but otherwise it's good to merge.
